### PR TITLE
Allow all mysql flags while importing database

### PR DESCRIPTION
**Pull request by abhijitrakas**
Commit for https://github.com/wp-cli/ideas/issues/112

### CODE REVIEW

File: src/DB_Command.php [hunk 1]
```diff
@@ -539,6 +539,9 @@ public function export( $args, $assoc_args ) {
 	 * [--dbpass=<value>]
 	 * : Password to pass to mysql. Defaults to DB_PASSWORD.
 	 *
+	 * [--<field>=<value>]
+	 * : Extra arguments to pass to mysqldump. [Refer to mysqldump docs](https://dev.mysql.com/doc/en/mysqldump.html#mysqldump-option-summary).
```
Comment (schlessera) [inline src/DB_Command.php]: We're not using `mysqldump` here, but rather the `mysql` binary. Comment and link to docs need to be adapted.

**Comment by abhijitrakas** [comment]
@schlessera Ok, cool. So we allow only flags which are mention in [MySql Doc](https://dev.mysql.com/doc/refman/8.0/en/mysqlimport.html) check `Table 4.12 mysqlimport Options` table. Correct me if I am wrong.

**Comment by schlessera** [comment]
@abhijitrakas No, we are actually using the `mysql` binary, not the `mysqlimport` one. So the options would be the ones documented here: https://dev.mysql.com/doc/refman/8.0/en/mysql-command-options.html

File: src/DB_Command.php [hunk 2]
```diff
@@ -1429,4 +1434,27 @@ private function get_colors( $assoc_args, $colors ) {
...
..., 'character-sets-dir', 'column-names', 'column-type-info', 'comments', 'compress', 'connect-expired-password', 'connect_timeout', 'database', 'debug', 'debug-check', 'debug-info', 'default-auth', 'default-character-set', 'defaults-extra-file', 'defaults-file', 'defaults-group-suffix', 'delimiter', 'enable-cleartext-plugin', 'execute', 'force', 'get-server-public-key', 'help', 'histignore', 'host', 'html', 'ignore-spaces', 'init-command', 'line-numbers', 'local-infile', 'login-path', 'max_allowed_packet', 'max_join_size', 'named-commands', 'net_buffer_length', 'no-beep', 'one-database', 'pager', 'pipe', 'plugin-dir', 'port', 'print-defaults', 'protocol', 'quick', 'raw', 'reconnect', 'i-am-a-dummy', 'safe-updates', 'secure-auth', 'select_limit', 'server-public-key-path', 'shared-memory-base-name', 'show-warnings', 'sigint-ignore', 'silent', 'skip-auto-rehash', 'skip-column-names', 'skip-line-numbers', 'skip-named-commands', 'skip-pager', 'skip-reconnect', 'socket', 'ssl-ca', 'ssl-capath', 'ssl-cert', 'ssl-cipher', 'ssl-crl', 'ssl-crlpath', 'ssl-fips-mode', 'ssl-key', 'ssl-mode', 'syslog', 'table', 'tee', 'tls-version', 'unbuffered', 'verbose', 'version', 'vertical', 'wait', 'xml' ];
```
Comment (schlessera) [inline src/DB_Command.php]: You should format this as a multiline array assignment, and use the plural for the variable name:
```suggestion
$allowed_mysql_options = [
			'auto-rehash',
			'auto-vertical-output',
			'batch',
			'binary-as-hex',
			'binary-mode',
			'bind-address',
			'character-sets-dir',
			'column-names',
			'column-type-info',
			'comments',
			'compress',
			'connect-expired-password',
			'connect_timeout',
			'database',
			'debug',
			'debug-check',
			'debug-info',
			'default-auth',
			'default-character-set',
			'defaults-extra-file',
			'defaults-file',
			'defaults-group-suffix',
			'delimiter',
			'enable-cleartext-plugin',
			'execute',
			'force',
			'get-server-public-key',
			'help',
			'histignore',
			'host',
			'html',
			'ignore-spaces',
			'init-command',
			'line-numbers',
			'local-infile',
			'login-path',
			'max_allowed_packet',
			'max_join_size',
			'named-commands',
			'net_buffer_length',
			'no-beep',
			'one-database',
			'pager',
			'pipe',
			'plugin-dir',
			'port',
			'print-defaults',
			'protocol',
			'quick',
			'raw',
			'reconnect',
			'i-am-a-dummy',
			'safe-updates',
			'secure-auth',
			'select_limit',
			'server-public-key-path',
			'shared-memory-base-name',
			'show-warnings',
			'sigint-ignore',
			'silent',
			'skip-auto-rehash',
			'skip-column-names',
			'skip-line-numbers',
			'skip-named-commands',
			'skip-pager',
			'skip-reconnect',
			'socket',
			'ssl-ca',
			'ssl-capath',
			'ssl-cert',
			'ssl-cipher',
			'ssl-crl',
			'ssl-crlpath',
			'ssl-fips-mode',
			'ssl-key',
			'ssl-mode',
			'syslog',
			'table',
			'tee',
			'tls-version',
			'unbuffered',
			'verbose',
			'version',
			'vertical',
			'wait',
			'xml',
		];
```

File: src/DB_Command.php [hunk 3]
```diff
@@ -1429,4 +1434,27 @@ private function get_colors( $assoc_args, $colors ) {
...
+
+		$mysql_args = array();
+
+		foreach ( $assoc_args as $mysql_option_key => $mysql_option_value ) {
+			// check flags is valid flag or not.
+			if ( in_array( $mysql_option_key,  $allowed_mysql_option, true ) && ! empty( $mysql_option_value ) ) {
```
Comment (schlessera) [inline src/DB_Command.php]: Use the plural for the allowed options array variable name:
```suggestion
			if ( in_array( $mysql_option_key,  $allowed_mysql_options, true ) && ! empty( $mysql_option_value ) ) {
```

File: src/DB_Command.php [hunk 4]
```diff
@@ -1429,4 +1434,27 @@ private function get_colors( $assoc_args, $colors ) {
...
+
+		$mysql_args = array();
+
+		foreach ( $assoc_args as $mysql_option_key => $mysql_option_value ) {
+			// check flags is valid flag or not.
```
Comment (schlessera) [inline src/DB_Command.php]: CS - Capitalize first word in comments:
```suggestion
			// Check flags to make sure they only contain valid options.
```

File: src/DB_Command.php [hunk 5]
```diff
@@ -1429,4 +1434,27 @@ private function get_colors( $assoc_args, $colors ) {
...
+			if ( in_array( $mysql_option_key,  $allowed_mysql_option, true ) && ! empty( $mysql_option_value ) ) {
+				$mysql_args[$mysql_option_key] = $mysql_option_value;
+
```
Comment (schlessera) [inline src/DB_Command.php]: Unneeded line break.
```suggestion
```

File: src/DB_Command.php [hunk 6]
```diff
@@ -1429,4 +1434,27 @@ private function get_colors( $assoc_args, $colors ) {
 
 		return $colors;
 	}
+
+	/**
+	 * Helper to pluck `mysql` opitons from associative args array.
```
Comment (schlessera) [inline src/DB_Command.php]: Typo:
```suggestion
	 * Helper to pluck `mysql` options from associative args array.
```

File: src/DB_Command.php [hunk 7]
```diff
@@ -1429,4 +1434,27 @@ private function get_colors( $assoc_args, $colors ) {
...
+	 *
+	 * @param array $assoc_args Associative args array.
+	 * @return array Array with `mysql` opitons set if in passed-in associative args array.
```
Comment (schlessera) [inline src/DB_Command.php]: Typo:
```suggestion
	 * @return array Array with `mysql` options set if in passed-in associative args array.
```

**Comment by schlessera** [comment]
Thanks for the PR, @abhijitrakas !

**Pull request merged**